File: 2016/03/03.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <array>


bool check_triangle (int, int, int);
std::array<int, 3> make_array (std::string string);


int main (void) {

    std::vector<std::array<int, 3>> problem_data;
    std::string temp_string;
    std::fstream file;

    file.open("./input.txt", std::ios::in);
    if (file.is_open()) {
        while (getline(file, temp_string)) {
            problem_data.push_back(make_array(temp_string));
        }
        file.close();
    }

    int sol_0 {}, sol_1 {};
    for (int i {}; i < problem_data.size(); i++) {
        sol_0 += check_triangle(problem_data[i][0], problem_data[i][1], problem_data[i][2]);
        if (i % 3 == 0) {
            sol_1 += 
                check_triangle(problem_data[i][0], problem_data[i+1][0], problem_data[i+2][0]) +
                check_triangle(problem_data[i][1], problem_data[i+1][1], problem_data[i+2][1]) +
                check_triangle(problem_data[i][2], problem_data[i+1][2], problem_data[i+2][2])
                ;
        }
    }

    std::cout << sol_0 << std::endl << sol_1 << std::endl;
}


bool check_triangle (int x, int y, int z) {
    return (x + y > z && y + z > x && z + x > y);
}


std::array<int, 3> make_array (std::string temp_string) {

    std::array<int, 3> array {};
    std::stringstream ssinput {temp_string};
    std::string segment {};

    for (int i {}; std::getline(ssinput, segment, ' ');) {
        if (segment != "") {
            array[i] = std::stoi(segment);
            i++;
        }
    }

    return array;
}

// 1.0 ms ± 0.6 ms (10k runs, 1k warmups) - Clang 14.0.6: clang++ -O3 -std=C++20

File: 2016/03/03.py
with open("./input.txt", 'r') as file:
    problem_data = [x.split() for x in file.readlines()]

def check_triangle(x, y, z):
    x, y, z = int(x), int(y), int(z)
    return x + y > z and y + z > x and z + x > y

sol_0, sol_1 = 0, 0
for i, row in enumerate(problem_data):
    sol_0 += check_triangle(*[int(x) for x in row])
    if i % 3 == 0:
        sol_1 += (
            check_triangle(problem_data[i][0], problem_data[i+1][0], problem_data[i+2][0]) +
            check_triangle(problem_data[i][1], problem_data[i+1][1], problem_data[i+2][1]) +
            check_triangle(problem_data[i][2], problem_data[i+1][2], problem_data[i+2][2])
        )

print(sol_0)
print(sol_1)

# 13.6 ms ±  0.9 ms (1k runs, 100 warmups) - Python 3.10.6 (conda-forge)